src/results.cpp

```diff
@@ -690,9 +693,21 @@ void Results::prepare_async()
     _impl::RealmCoordinator::register_notifier(m_notifier);
 }
 
-NotificationToken Results::add_notification_callback(CollectionChangeCallback cb) &
+#if REALM_ENABLE_SYNC
+NotificationToken Results::add_notification_callback(CollectionChangeCallback cb, util::Optional<std::string> subscription_name) &
+#else 
+NotificationToken Results::add_notification_callback(CollectionChangeCallback cb, util::Optional<std::string>) &
+#endif
 {
     prepare_async();
+#if REALM_ENABLE_SYNC
+    if (m_realm->is_partial() && !m_have_subscribed) {
```

Comment:
Subscribing to the query needs to be explicit, and separate from simply adding a notification callback, in order to support the API that Cocoa and .NET are exposing (maybe also JS?). It seems like having it be explicit at this level shouldn't be problematic for Java since you can simply call the `subscribe` method (or whatever it's called) at the same point you call `add_notification_callback`.

Reply:
Yes, we can easily do that.

Reply:
It's worth noting that using the notification callback to deliver changes in partial sync state introduces the artificial limitation of it being impossible to subscribe to a partial sync query while inside a write transaction.